
################################################################################
#    &&&....&&&    % Project: MSPJ approach for identification of DEGs         #
#  &&&&&&..&&&&&&  % Author: Bo Li, Huachun Yin, Jingxin Tao, Youjin Hao       #
#  &&&&&&&&&&&&&&  % Date: Jun. 1st, 2020                                      #
#   &&&&&&&&&&&&   %                                                           #
#     &&&&&&&&     % Environment: R version 3.5.3;                             #
#       &&&&       % Platform: x86_64-pc-linux-gnu (64-bit)                    #
#        &         %                                                           #
################################################################################

### ****************************************************************************
### code chunk number 03: The meta-analysis approach based on resampling.
### ****************************************************************************

### Title: A novel meta-analytical approach to improve identification of DEGs. 

# input: eset - a gene expression matrix, genes in rows and samples in columns. 
#        set.n - the number of the subgroups generated by sampling. 
# output: degs - a genelist, including the gene names, pooled SMDs, and so on. 

### ------------------------------------------------------------------------ ###
### Step-01. Loading the gene expression matrix in *.csv, or *.RData format. 

# seq.matrix.RData - simulated RNA-seq data; 
# mcr.matrix.RData - simulated microarray data. 

# Taking the microarray data as example. 

eset <- get(load("mcr.matrix.RData")); rm(mcr.matrix)
# eset <- get(load("seq.matrix.RData")); rm(seq.matrix)

# DT::datatable(eset)

### End of Step-01.
### ------------------------------------------------------------------------ ###

### ------------------------------------------------------------------------ ###
### Step-02. Setting the parameters used for re-sampling.

# set.n: the times of resampling, or the number of sub-groups.   
# size.min: the lower limit of sample size in each group.
# size.max: the maximum sample size in each group. 
# ord.gene: which gene you focused on. 

#. set.n <- 40
#. size.min <- 10
#. size.max <- 20
#. ord.gene <- 10

### ------------------------------------------------------------------------ ###
### Step-03. Generating multiple sub-groups based resampling for primary study. 

sample.sets <- generateSubGroup(eset, set.n = 40, size.min = 10, size.max = 20)

# dim(subgroups[[5]])

### End of Step-03.
### ------------------------------------------------------------------------ ###

### ------------------------------------------------------------------------ ###
### Step-04. Computing the statistics used for meta-analysis.

set.n <- length(sample.sets)

num.gene <- nrow(eset)

batchMeta(data.list = sample.sets, 
          cutoff = 0.5, 
          g.start = 100, 
          g.end = 500)



#. MetaSet <- list()




# Funnel plot for a given gene (ord.gene). 

col.seq <- rep(NA, ncol(eset))

col.seq[grep("Experimental", colnames(eset))] <- "red"
col.seq[grep("Control", colnames(eset))] <- "green"

funnel(res, 
       pch = "S", 
       col = col.seq, 
       bg = 1:40)

# End. 


