
################################################################################
#    &&&....&&&    % Project: MSPJ approach for identification of DEGs         #
#  &&&&&&..&&&&&&  % Author: Bo Li, Huachun Yin, Jingxin Tao, Youjin Hao       #
#  &&&&&&&&&&&&&&  % Date: Jun. 1st, 2020                                      #
#   &&&&&&&&&&&&   %                                                           #
#     &&&&&&&&     % Environment: R version 3.5.3;                             #
#       &&&&       % Platform: x86_64-pc-linux-gnu (64-bit)                    #
#        &         %                                                           #
################################################################################

### ****************************************************************************
### code chunk number 03: A nested meta-analysis approach based on resampling.
### ****************************************************************************

### Title: A novel meta-analytical approach to improve identification of DEGs. 

# input: eset - a gene expression matrix, genes in lines and samples in columns. 
#        set.n - the number of the subgroups generated by sampling. 
# output: degs - a genelist, including the gene names, pooled SMDs, and so on. 

### ------------------------------------------------------------------------ ###
### Step-01. Loading the gene expression data in *.csv

load("seq.matrix.RData")

library(meta)

set.n <- 40

size.min <- 10

size.max <- 20

ord.gene <- 10

### ****************************************************************************
### code chunk number 01: Loading the gene expression data in *.csv
### ****************************************************************************

### ------------------------------------------------------------------------ ###
### Step-01. Preparing the GEO accession number for GEO Series.

eset <- as.data.frame(matrix(rnorm(20000 * 20, 10, 5), 20000, 20))

eset[1:100, 1:8] <- eset[1:100, 1:8] + 50

eset[101:200, 9:20] <- eset[101:200, 9:20] + 50

rownames(eset) <- paste("Gene", 1:20000, sep = "-")

colnames(eset) <- c(paste("Experimental", 1:8, sep = "-"), 
                    paste("Control", 1:12, sep = "-"))

sample.sets <- list()

for (i in 1:set.n) {
  
  sample.sets[[i]] <- eset[, sample(1:ncol(eset), sample(size.min:size.max, 1, replace = FALSE), 
                                    replace = TRUE)]
  
}

names(sample.sets) <- paste("sampling_set", 1:set.n, sep = "-")

### ------------------------------------------------------------------------ ###
### Step-02. Computing the statistics used for meta-analysis.

stat.mat <- data.frame(matrix(NA, set.n, 8))

names(stat.mat) <- c("study", "year", 
                     "n.e", "mean.e", "sd.e", 
                     "n.c", "mean.c", "sd.c")

stat.mat$study <- paste("sampling_set", 1:set.n, sep = "-")

stat.mat$year <- sample(2000:2020, set.n, replace = TRUE)

# x <- sample.sets[[1]]

f.ne <- function(x) length(grep("Experimental", colnames(x)))

f.meane <- function(x) apply(x[ord.gene, grep("Experimental", colnames(x))], 1, mean)

f.sde <- function(x) apply(x[ord.gene, grep("Experimental", colnames(x))], 1, sd)

f.nc <- function(x) length(grep("Control", colnames(x)))

f.meanc <- function(x) apply(x[ord.gene, grep("Control", colnames(x))], 1, mean)

f.sdc <- function(x) apply(x[ord.gene, grep("Control", colnames(x))], 1, sd)

stat.mat$n.e <- unlist(lapply(sample.sets, f.ne))
stat.mat$n.c <- unlist(lapply(sample.sets, f.nc))

stat.mat$mean.e <- unlist(lapply(sample.sets, f.meane))
stat.mat$mean.c <- unlist(lapply(sample.sets, f.meanc))

stat.mat$sd.e <- unlist(lapply(sample.sets, f.sde))
stat.mat$sd.c <- unlist(lapply(sample.sets, f.sdc))




### ------------------------------------------------------------------------ ###
### Step-01. Preparing the GEO accession number for GEO Series.

# load("Rimage_6.11.RData")

# My Mac OS. 
# setwd("/Users/libo/Test/Asthma/AsthmaData/")



# data(Fleiss93cont)

res <- metacont(n.e, # Number of observations in experimental group
               mean.e, # Estimated mean in experimental group
               sd.e, # Standard deviation in experimental group
               n.c, # Number of observations in control group
               mean.c, # Estimated mean in control group
               sd.c, # Standard deviation in control group
               studlab = study, # An optional vector with study labels
               data = stat.mat, # Data frame containing the study information
               sm = "SMD")  # One of three measures ("MD", "SMD" and "ROM")

forest(res, 
       col.fix = "red", 
       col.random = "blue",
       col.study = "black",
       col.square = "gray",
       #. col.square.lines = col.square,
       col.inside = "white",
       col.diamond = "gray",
       #. col.diamond.fixed = col.diamond,
       #. col.diamond.random = col.diamond,
       col.diamond.lines = "black",
       #. col.diamond.lines.fixed = col.diamond.lines,
       #. col.diamond.lines.random = col.diamond.lines,
       #. col.inside.fixed = col.inside,
       #. col.inside.random = col.inside,
       col.predict = "red",
       col.predict.lines = "black",
       col.by = "darkgray",
       col.label.right = "black",
       col.label.left = "black")


funnel(res, 
       pch = 24, 
       col = 1:40, 
       bg = 1:40)

#. str(res)
#. 
#. if (res$overall.hetstat) {
#.   
#.   cat("The heterogeneity of this model is satisfactory")
#.   
#.   if (abs(res$TE.random) > 0.2  & abs(res$TE.fixed) > 0.2) {
#.     
#.     cat("\nThis gene is a differentially expressed gene")
#.     
#.     if (res$TE.random > 0.2) cat("\nState: up-regulated") else cat("\nState: down-regulated")
#.   }  
#.   
#. }
#. 
#. 
#. res$TE.random

