
################################################################################
#    &&&....&&&    % Project: MSPJ approach for identification of DEGs         #
#  &&&&&&..&&&&&&  % Author: Bo Li, Huachun Yin, Jingxin Tao, Youjin Hao       #
#  &&&&&&&&&&&&&&  % Date: Jun. 1st, 2020                                      #
#   &&&&&&&&&&&&   %                                                           #
#     &&&&&&&&     % Environment: R version 3.5.3;                             #
#       &&&&       % Platform: x86_64-pc-linux-gnu (64-bit)                    #
#        &         %                                                           #
################################################################################

### ****************************************************************************
### code chunk number 05: Analysis of DEGs based permutation test.
### ****************************************************************************

### Reference: Implementing a class of permutation tests: the coin package, 2008. 
### Ref: https://rcompanion.org/handbook/K_01.html

# input: eset - a gene expression matrix, genes in lines and samples in columns. 
#        set.n - the number of the subgroups generated by sampling. 
# output: degs - a genelist, including the gene names, pooled SMDs, and so on. 

### ------------------------------------------------------------------------ ###
### Step-01. Loading the gene expression data in *.csv

input <- get(load("input.RData"))

### End of Step-01.
### ------------------------------------------------------------------------ ###

### ------------------------------------------------------------------------ ###
### Step-02. Identifying the DEGs by permutation test, and p < 0.05.

# using coin package. 
# independence_test(y ~ tr, alternative = "greater")  # one-tailed
# independence_test(y ~ tr, alternative = "two.sided")  # two-tailed

library(FSA)

all.genes <- names(input)[-1]

gene.count <- length(all.genes)

deg.count <- NULL

i <- 0

pb <- txtProgressBar(min = 0, max = gene.count, style = 3, char = "+")

for (g in all.genes) {
  
  i <- i + 1
  
  setTxtProgressBar(pb, i)
  
  # Display the progress bar!

  tmp <- Summarize(get(g) ~ sam.lab, data = input, digits = 3)
  
  # print(tmp)
  # boxplot(get(g) ~ sam.lab, data = input)
  
  deg.per <- try(independence_test(get(g) ~ sam.lab, 
                                   data = input), 
                 silent = FALSE)
  
  # deg.Z <- deg.per@statistic@teststatistic
  deg.p <- deg.per@distribution@pvalue(deg.per@statistic@teststatistic)
  
  print(i)
  
  flush.console()
  
  # This variable, deg.count, stores all the differentially expressed genes.
  
  if (!is.na(deg.p) & deg.p < 0.01) deg.count <- c(deg.count, g) else next
  
}

close(pb)

deg.per <- deg.count; rm(deg.count)

deg.per

# For permutation test of independence
# For two groups as independent samples, and tests if there is a difference in values between the two groups.

# For permutation test of symmetry. 
# For two groups as having paired or repeated data, paired within Individual.

# deg.per <- symmetry_test(g10000 ~ sam.lab | Individual, data = input)

#. class(deg.per)
#. getMethod("show","ScalarIndependenceTest")

### End of Step-02.
### ------------------------------------------------------------------------ ###

### ------------------------------------------------------------------------ ###
### Step-03. Identifying the DEGs by Student t-test and Fold change methods.

fc <- NULL
p <- NULL
for (i in 2:ncol(input)) {
  
  print(i)
  x <- mean(input[, i][input$sam.lab == "Experimental"])/mean(input[, i][input$sam.lab == "Control"])
  
  y <- wilcox.test(input[, i][input$sam.lab == "Experimental"], input[, i][input$sam.lab == "Control"], 
              paired = FALSE) # paired = TRUE indicates that, wilcoxon sign-rank test. 
  
  fc <- c(fc, x)
  
  p <- c(p, y$p.value)
  
  # Sys.sleep(3)
}

plot(fc, -log(p, 10), xlim = c(0, 10))

abline(h = -log(0.05, 10), col = "red")

abline(v = c(0.5, 2), col = "blue")

deg.wilcox <- colnames(input)[-1][((fc < 0.5 | fc > 2) & p < 0.05)]

hist(p, breaks = 100)

table(p[1:500] < 0.05)

hist(input$g1, breaks = 10)

### End of Step-03.
### ------------------------------------------------------------------------ ###
