
################################################################################
#    &&&....&&&    % Project: MSPJ approach for identification of DEGs         #
#  &&&&&&..&&&&&&  % Author: Bo Li, Huachun Yin, Jingxin Tao   
#  &&&&&&&&&&&&&&  % Date: Mar. 1st, 2022                                      #
#   &&&&&&&&&&&&   %                                                           #
#     &&&&&&&&     % Environment: R version 3.5.3;                             #
#       &&&&       % Platform: x86_64-pc-linux-gnu (64-bit)                    #
#        &         %                                                           #
################################################################################

### ****************************************************************************
### code chunk number 06: Analysis of DEGs based permutation test.
### ****************************************************************************

### Reference: Implementing a class of permutation tests: the coin package, 2008. 
### Ref: https://rcompanion.org/handbook/K_01.html

# input: eset - a gene expression matrix, genes in lines and samples in columns. 
#        set.n - the number of the subgroups generated by sampling. 
# output: degs - a genelist, including the gene names, pooled SMDs, and so on. 

### ------------------------------------------------------------------------ ###
### Step-01. Loading the gene expression data in *.csv

input <- get(load("input.RData"))

print(input[1:6, 1:6])

### End of Step-01.
### ------------------------------------------------------------------------ ###

### ------------------------------------------------------------------------ ###
### Step-02. Identifying the DEGs by permutation test, and p < 0.05.

# using coin package. 
# independence_test(y ~ tr, alternative = "greater")  # one-tailed
# independence_test(y ~ tr, alternative = "two.sided")  # two-tailed
pvalue<-0.05

#set the cutoff of p value

library(FSA)
all.genes <- names(input)[-1]
gene.count <- length(all.genes)
deg.count <- NULL

i <- 0

pb <- txtProgressBar(min = 0, max = gene.count, style = 3, char = "+")

for (g in all.genes) {
  
  i <- i + 1
  
  setTxtProgressBar(pb, i)
  
  # Display the progress bar!
  
  tmp <- Summarize(get(g) ~ sam.lab, data = input, digits = 3)
  
  #print(tmp)
  #boxplot(get(g) ~ sam.lab, data = input)
  
  deg.per <- try(independence_test(get(g) ~ sam.lab, 
                                   data = input), 
                 silent = FALSE)
  
  # deg.Z <- deg.per@statistic@teststatistic
  deg.p <- deg.per@distribution@pvalue(deg.per@statistic@teststatistic)
  
  flush.console()
  
  # This variable, deg.count, stores all the differentially expressed genes.
  
  if (!is.na(deg.p) & deg.p < pvalue) deg.count <- c(deg.count, g,deg.p) else next
  
}

close(pb)


deg.count1<-matrix(deg.count,ncol=2,byrow = T)

colnames(deg.count1)<-c("name","Pvalue")

deg.count1<-as.data.frame(deg.count1)


fdrp=p.adjust(deg.count1$Pvalue, "BH")


names(fdrp)<-deg.count1$name

deg.count2<-as.matrix(fdrp)

colnames(deg.count2)<-"Pvalue"

deg.count2<-as.data.frame(deg.count2)

deg.count2$name<-rownames(deg.count2)

deg.count2 <- deg.count2[order(deg.count2$Pvalue),]

deg.per <-deg.count2[deg.count2$Pvalue<pvalue,]

deg.per <- deg.count1$name

deg.per

# For permutation test of independence
# For two groups as independent samples, and tests if there is a difference in values between the two groups.

# For permutation test of symmetry. 
# For two groups as having paired or repeated data, paired within Individual.

# deg.per <- symmetry_test(g10000 ~ sam.lab | Individual, data = input)

#. class(deg.per)
#. getMethod("show","ScalarIndependenceTest")

### End of Step-02.
### ------------------------------------------------------------------------ ###
