
################################################################################
#    &&&....&&&    % Project: MSPJ approach for identification of DEGs         #
#  &&&&&&..&&&&&&  % Author: Bo Li, Huachun Yin, Jingxin Tao, Youjin Hao       #
#  &&&&&&&&&&&&&&  % Date: Jun. 1st, 2020                                      #
#   &&&&&&&&&&&&   %                                                           #
#     &&&&&&&&     % Environment: R version 3.5.3;                             #
#       &&&&       % Platform: x86_64-pc-linux-gnu (64-bit)                    #
#        &         %                                                           #
################################################################################

### ****************************************************************************
### code chunk number 03: The meta-analysis approach based on resampling.
### ****************************************************************************

### Title: A novel meta-analytical approach to improve identification of DEGs. 

# input: eset - a gene expression matrix, genes in rows and samples in columns. 
#        set.n - the number of the subgroups generated by sampling. 
# output: degs - a genelist, including the gene names, pooled SMDs, and so on. 

### ------------------------------------------------------------------------ ###
### Step-01. Computing the statistics used for meta-analysis.

set.n <- length(sample.sets)

num.gene <- nrow(eset)

# 1) Identifying the differentially expressed genes by meta-analysis, in bulk. 

deg <- batchMeta(data.list = sample.sets, 
                 cutoff = 0.5, 
                 g.start = 1, 
                 g.end = num.gene)

deg.meta <- c(rownames(eset)[deg$UpGene], 
              rownames(eset)[deg$DownGene])

# 2) Identifying the differentially expressed genes by meta-analysis, one by one. 

# -- Generate the meta object. 

res <- singleMeta(data.list = sample.sets, 
                  ord.gene = 10)

# -- Draw the forest plot.  

forest(res, 
       col.fix = "red", 
       col.random = "blue",
       col.study = "black",
       col.square = "gray",
       #. col.square.lines = col.square,
       col.inside = "white",
       col.diamond = "gray",
       #. col.diamond.fixed = col.diamond,
       #. col.diamond.random = col.diamond,
       col.diamond.lines = "black",
       #. col.diamond.lines.fixed = col.diamond.lines,
       #. col.diamond.lines.random = col.diamond.lines,
       #. col.inside.fixed = col.inside,
       #. col.inside.random = col.inside,
       col.predict = "red",
       col.predict.lines = "black",
       col.by = "darkgray",
       col.label.right = "black",
       col.label.left = "black")

# -- Draw the funnel plot for a given gene. 

col.seq <- rep(NA, ncol(eset))

col.seq[grep("Experimental", colnames(eset))] <- "red"
col.seq[grep("Control", colnames(eset))] <- "green"

funnel(res, 
       pch = 20, 
       col = col.seq, 
       bg = 1:50)

### End of Step-01.
### ------------------------------------------------------------------------ ###
